### ISSUE TITLE
Meshagent _MSH prefers embeded MSH over external

### ISSUE THREAD

**Issue by HuFlungDu** [Ylianst/MeshAgent#258]
I'm not actually sure this is a bug or working as intended. It seems counter intuitive to me, but feel free to close if this is working as intended.

`ILibDuktape_Polyfills_MSH` function seems to have a preference for loading the MSH values embedded in the executable over ones in the MSH file. This makes the MSH console function behave in an unexpected way. `msh set` will write to the MSH file, but then `msh` will return the data in the executable, rather than from the msh file. Confusingly, on the C side, `importSettings` within `MeshAgent_AgentMode` will override `checkForEmbeddedMSH` with the values in the .msh file, so in C, you will reference the settings in the msh file, while if you call `_MSH()` in javascript, you'll get the embedded values.

This only seems to effect windows agents (though I haven't tested MacOS), presumably because linux agents store the msh values differently, but it causes the agents to act inconsistently on different OSs.

I believe fixing this would require doing some merging of the objects, and probably parsing both the embedded and the msh file, overriding with the MSH when necessary.

**Comment by si458** [comment]
The new `msh set` function i added does say u need to restart the meshagebt for it to take effect

Are u sure u are restarting the meshagent after u set things with `msh set`? Then checking `msh get` again?

**Comment by HuFlungDu** [comment]
Yep. I noticed the issue because I was trying to modify the .msh file while the agent was down to add a new option and it doesn't get picked up. Just checked it again just in case and yeah, it still doesn't pick it up in the case I mentioned.

**Comment by si458** [comment]
what msh value are you trying to change to take effect inside the agent as it might be hard coded to ignore the msh

**Comment by HuFlungDu** [comment]
I'm adding a new key called "validateTLS" for a security update I'm working on. I can set it when I download the agent (in the meshcentral server code where it fills out the embedded values), and it is respected by the javascript in that case, but if I change it (either using the msh command or modifyng the file directly) the C code respects my change, but the javascript code does not. Having it cleared from the downloaded agent also works, but then I can't set it. It's set in stone to the value it was set during agent download.

**Comment by si458** [comment]
oh yeh sorry i just remembered! 🤦 
https://github.com/Ylianst/MeshCentral/issues/5845#issuecomment-2033262310
calling `msh` reads the value from its memory
where as `msh get myvalue` gets the value from your msh file!

**Comment by HuFlungDu** [comment]
Nope:

```
> msh set test "test"
msh set test successful
> msh
{
  "MeshName": "Test",
  "MeshType": "2",
  "MeshID": "xxxxxxx",
  "ServerID": "xxxxxxx",
  "MeshServer": "wss://localhost:443/agent.ashx",
  "translation": ...
}
> msh get "test"
Unknown Value: test
```

And my MSH file:
```
MeshName=Test
MeshType=2
MeshID=xxxxxxx
ServerID=xxxxxxx
MeshServer=wss://localhost:443/agent.ashx
translation=...
test=test
```

Tested after restarting and got the same issue.



Which makes sense; they both just call `_MSH()`, then `get` just pulls the value out of the object that returns, so they wouldn't return different results.

For me, this is a functionality problem; if I set my new "validateTLS" option, then I check it by calleng `_MSH()` in the JS code, it will still return false, even though the user set it to true. I will never be able to change it after the file is downloaded. Additionally, this acts differently on Linux than it does on Windows, so I suspect one of them is incorrect.

I can fix the issue, I just need to know what the resolution should be:

Should embedded take priority over external? If so, we need to change the C code so that it prefers embedded.
Should embedded be ignored if external is found? That's an easy fix of just moving a line a little further down and swapping a comparison.
Should embedded be the default and then individual values be overridable by external? Then we'll need to do the parsing twice (or maybe just append the external after the embedded, then parse. Looks like that should work since the parsing is ordered).

**Comment by si458** [comment]
Huh? That's weird? It use to work and all worked when I added it and tested it!?
What os/version is the meshagent running on?

**Comment by HuFlungDu** [comment]
Windows only. Works as I would expect on Linux.

I'm using the agent pulled from commit 61fb6898c0424b29bee5c91a1faa350d85e5b741 on the meshcentral repo for my test server. My production server is using meshcentral version 1.1.30. Both are exhibiting this issue.

meshagent -info:
```
Agent.exe -info
Compiled on: 12:12:40, Dec  9 2022
   Commit Hash: 1ac56a2aa26cc403efe758fc53386a3dc0e706b7
   Commit Date: 2022-Dec-2 11:42:16-0800
Using OpenSSL 1.1.1s  1 Nov 2022
Agent ARCHID: 4
Detected OS: Microsoft Windows 11 Pro - 23H2/22631 - x64
LOCKED to: xxxx.xxxx.com
 => xxxxxxx
```

**Comment by si458** [comment]
Yeh, that's bloody weird, as I just tried it here, and it's not working either!
It sets the value in the file but doesn't return it?
Even if u set a value that use haven't embedded like a completely new unused one,
Even restarting the agent too doesn't load the value, which is weird?

**Comment by HuFlungDu** [comment]
It's caused by this code:

https://github.com/Ylianst/MeshAgent/blob/90f730c3d49555e5c16106a7c8dd0219818edad6/microscript/ILibDuktape_Polyfills.c#L3345-L3346

If there is any embedded MSH, it will skip loading the msh file. This might be one of them Schroedinbugs, lol. Looks like it works on Linux because the Linux executable includes an embedded javascript file which contains the MSH as a POJO, which `checkForEmbeddedMSH_ex` was not coded to handle, while the windows executable includes the MSH as a basic string at the end of the file, which is what `checkForEmbeddedMSH_ex` looks for.

**Comment by si458** [comment]
so ive just got in today to have alook at this, and tested a random windows 10 pro machine and its working?
this is very weird?
```
> msh
{
  "MeshName": "myothergroup",
  "MeshType": "2",
  "MeshID": "LONGASSMESHIDHERE",
  "ServerID": "LONGASSSERVERIDHERE",
  "MeshServer": "wss://mc.myserver.com:443/agent.ashx",
  "displayName": "MyRemoteAgent",
  "description": "MyRemoteAgentDescription",
  "companyName": "MyCompanyName",
  "meshServiceName": "MyRemoteAgent",
  "fileName": "myremoteagent",
  "translation": "LONGASSTRANSLATEJSONSTRINGHERE",
  "installedByUser": "S-1-5-21-2932858411-1413530269-3412965231-1001"
}
> msh help
Proper usage: msh [get|set|delete]
msh get MeshServer
msh set abc "xyz"
msh delete abc
> msh set bollox "abc"
msh set bollox successful
> msh get bollox
abc
> msh
{
  "MeshName": "myothergroup",
  "MeshType": "2",
  "MeshID": "LONGASSMESHIDHERE",
  "ServerID": "LONGASSSERVERIDHERE",
  "MeshServer": "wss://mc.myserver.com:443/agent.ashx",
  "displayName": "MyRemoteAgent",
  "description": "MyRemoteAgentDescription",
  "companyName": "MyCompanyName",
  "meshServiceName": "MyRemoteAgent",
  "fileName": "myremoteagent",
  "translation": "LONGASSTRANSLATEJSONSTRINGHERE",
  "installedByUser": "S-1-5-21-2932858411-1413530269-3412965231-1001",
  "bollox": "abc"
}
> msh                <-   i changed the .msh file before running this command
{
  "MeshName": "myothergroup",
  "MeshType": "2",
  "MeshID": "LONGASSMESHIDHERE",
  "ServerID": "LONGASSSERVERIDHERE",
  "MeshServer": "wss://mc.myserver.com:443/agent.ashx",
  "displayName": "MyRemoteAgent",
  "description": "MyRemoteAgentDescription",
  "companyName": "MyCompanyName",
  "meshServiceName": "MyRemoteAgent",
  "fileName": "myremoteagent",
  "translation": "LONGASSTRANSLATEJSONSTRINGHERE",
  "installedByUser": "S-1-5-21-2932858411-1413530269-3412965231-1001",
  "bollox": "xyz"
}
```

**Comment by si458** [comment]
so it appears windows 11 machines dont work as ive just tried a few machines
but windows 10 machines work no problem!
so the indeed might be a bug with the underlying code of meshagent then

**Comment by si458** [comment]
ok so now my windows 11 devices are working?
all i did was restart the pcs and then its working?
im wondering if this a memory leak maybe?

**Comment by si458** [comment]
@HuFlungDu just a strange one,
are you sure when you run `msh set test "test"` it says `successful` and not `unsuccessful` ?
as im spotting on one machine im getting `unsuccessful` and this only happens if it cant write to the `.msh` file?
```
> msh set test "test"
msh set test unsuccessful
```

**Comment by HuFlungDu** [comment]
Yeah, I just copied from the mesh console and redacted for my post above; if it had said unsuccessful it would be up there. I can also confirm that the MSH file IS edited correctly with the MSH command.

On the agents where it is working, if you run `powershell -command "& {Get-Content .\agent.exe -last 5}"`, does it look like there is an embedded msh file? I'm not sure the cases in which an embedded mesh is added, maybe some aren't getting it.

**Comment by HuFlungDu** [comment]
@Ylianst Is this intended so that an executable can be signed and then this MSH cannot be modified? You mentioned that in the community meeting, checking if that's what's going on here.
